File: mobile/src/hiddenApiStubs/java/android/net/wifi/SoftApInfo.java
package android.net.wifi;

import android.net.MacAddress;
import android.os.Parcelable;
import androidx.annotation.RequiresApi;

import java.util.List;

@RequiresApi(30)
public abstract class SoftApInfo implements Parcelable {
    @RequiresApi(33)
    public static final int CHANNEL_WIDTH_AUTO = -1;

    /**
     * Currently only used on API 30.
     */
    public static final int CHANNEL_WIDTH_INVALID = 0;

    @RequiresApi(31)
    public abstract long getAutoShutdownTimeoutMillis();

    public abstract int getBandwidth();

    @RequiresApi(31)
    public abstract MacAddress getBssid();

    public abstract int getFrequency();

    /**
     * Used on API 30+ when present; missing implementations are tolerated below API 36.
     */
    public abstract MacAddress getMldAddress();

    @RequiresApi(31)
    public abstract String getApInstanceIdentifier();

    @RequiresApi(35)
    public abstract List<OuiKeyedData> getVendorData();

    @RequiresApi(31)
    public abstract int getWifiStandard();
}

File: mobile/src/main/java/be/mygod/vpnhotspot/net/wifi/SoftApInfo.kt
package be.mygod.vpnhotspot.net.wifi

import android.net.wifi.SoftApInfo
import android.os.Build
import androidx.annotation.RequiresApi
import be.mygod.vpnhotspot.util.ConstantLookup
import be.mygod.vpnhotspot.util.UnblockCentral
import timber.log.Timber

@RequiresApi(30)
val softApChannelWidthLookup = ConstantLookup("CHANNEL_WIDTH_") { SoftApInfo::class.java }

@get:RequiresApi(31)
val SoftApInfo.apInstanceIdentifierOrNull: String? get() = try {
    UnblockCentral.getApInstanceIdentifier(this)
} catch (e: NoSuchMethodError) {
    if (Build.VERSION.SDK_INT >= 31) Timber.w(e)
    null
}
